MCC-2291 Readme updates (#823)
mobilecoinfoundation/mobilecoin: attest/README.md

Remote attestation support. This family of crates provides support for Intel's EPID remote attestation in MobileCoin enclaves.

# Crates

| Name | Purpose |
| ---- | ------- |
| `ake` | Authenticated key exchange using the noise framework |
| `api` | gRPC Attestation APIs and protobufs |
| `core` | EPID remote attestation types and agnostic primitives |
| `enclave-api` | Serializeable structures for common enclave functions |
| `net` | IAS API Client |
| `trusted` | Traits used to monkey-patch enclave-only functionality onto `core` types |
| `untrusted` | Traits used to monkey-patch untrusted-only functionality onto `core` types |

# Enclave Concepts and Jargon

Intel's enclaves and EPID-based remote attestation is a very complex subject (there are many layers and protocols and ways to stitch things together), so the first thing to get out of the way is understanding the basics and the jargon. We'll use the MobileCoin validator (the "consensus service") for this example.

## Trusted Computing Base

> ***Trusted Computing Base*** (TCB) is pure jargon. It describes the much more pragmatic "all the things you need to believe haven't been pwned (things you trust) in order to believe the system hasn't been pwned (can be trusted)." It naturally follows that one great way to improve security is to reduce the number of things that you have to believe in to be however secure you need to be.

Reviewer: ```suggestion
***Trusted Computing Base*** (TCB) is jargon for "all the things you need to trust that haven't been compromised in order to believe the system can be trusted."
```

> To show how a TCB comes into being, lets take your typical "take a JSON document from HTTP, process it, return more JSON" web application, for example, isn't going to be passing raw network card buffers in RAM around from function to function. Instead you're going to have the OS deserialize the ethernet frame (and IP packet, and TCP segment), a TLS library decrypt and verify the encrypted data, a web server deserialize the HTTP request, and a framework/parser deserialize the JSON into an object, all before your application ever sees anything.

Reviewer: ```suggestion
To show how a TCB comes into being, take a typical scenario of a web application receiving a JSON document via HTTP, processing it and returning a JSON response. Encapsulation makes this seem like a single process, but under the hood, the OS deserializes the ethernet frame/IP packet/TCP segment, a TLS library decrypts and verifies the encrypted data, a web server deserializes the HTTP request, and a framework/parser deserializes the JSON into an object before the application sees the original message.
```

> And ultimately, you're going to create a TCB simply by following good development practices. Stabilizing a procedure and including it in a library that is shared in multiple places makes that library part of your TCB. You've established an ad-hoc trust in that procedure because you have used it and you "know" it works, you have unit tests around it to ensure you exercise edge cases, etc.

Reviewer: ```suggestion
A TCB is created following good development practices. Stabilizing a procedure and including it in a library that is shared in multiple places makes that library part of a TCB. In doing so, ad-hoc trust in that procedure is established because it's been used extensively, tested to exercise edge cases, etc.
```

> All that testing and ad-hoc trust aside, there's an even deeper analogue to be learned. Ultimately, you want to be in the business of squashing *classes* of security bugs. It's all fine and good to fix a SQL injection bug or a string misquote, but what you really want to do is make your code mis-use resistant, so SQL injection or string misquoting bugs are simply impossible to write.

Reviewer: ```suggestion
Outside of informal trust, it is important that you're protecting against *classes* of security bugs. Code should be resistant to mis-use. For example, in a hypothetical system relying on SQL, SQL injection vulnerabilities shouldn't needed to be patched, such vulnerabilities should be impossible to write in the first place.
```

> Another way to get to the same place, which seems much weaker but actually has a richer *history* is to make bugs like those *not security impacting*. That is, OK, fine, so there's a SQL injection. If you are talking to a database with read-only credentials with restricted query times, that's not great, but it's also not that bad.

Reviewer: ```suggestion
```

## Attestation

This "telling the difference" is handled by attestation, which is a multi-layered technique using building blocks from lower levels.

1. Local attestation (most of the actual work of attestation is done locally, on the host)
1. Quoting enclaves (by passing data to the quoting enclave)
1. IAS API call (a REST API call to IAS to sign off on the local attestation data)
1. Verification (verifying the result from IAS is what we want)

### Between Enclaves on the same CPU

> Don't skip this section just because you're interested in remote attestation. It's a core building block if you want to understand how the system works.

The first thing to know about attestation is that it begins with Alice The Enclave asking the CPU for it's own `EREPORT`. The CPU will fill in a data structure containing information about Alice.

> The contents of Alice's `EREPORT` can be used to construct a `TARGETINFO` data structure, which is given to Bob The Enclave. Bob Alice's targetting information when he calls `EREPORT` from within his context, and he CPU fills in the report as before, and adds an authentication code so Alice can verify Bob's report is valid.

Reviewer 2: Bob Alice's?

> Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to here and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.

Reviewer 2: ```suggestion
Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to her and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.
```

> Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to her and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.

Author: I agree with EREPORT (the command) vs. REPORT (the structure), but I'm going to keep it simple, there's no need to discuss EGETKEY (which is never directly called in any of our code).

> Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to her and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.

Reviewer 3: Ok, but I think the 'message' concept is important. Attestation is meaningless without the report message.

> Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to her and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.

Author: I disagree, not because `REPORTDATA` isn't the message (in a cryptographic sense), but because it should not be treated as a "message" by an author of an application that wants to use an enclave. It may be interesting to know, but it isn't necessary to understanding, and will lead folks down the wrong path.

For example: I think one mistake we made was putting the block signing key into the AVR. Now, if we want to save round-trips by adding more details to authentication (e.g. adding the configured minimum fee to that would save a round-trip during tx construction, and an enclave verifier could check that the fees match directly, and provide better error reporting, rather than stuffing it into the noise prologue), we're kind of screwed. If we'd treated those bytes as an authentication tag for a larger data structure instead (i.e. as a bootstrap for a *real* message), I think we'd be in a better spot.

> Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to her and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.

Reviewer 3: I don't really understand. Isn't proving a piece of data (message) was made by an enclave the entire raison d'etre of attestation? You don't gain anything from a `REPORT` or AVR that has no message.

### Quoting Enclaves

For remote attestation purposes, Alice is *Intel's Quoting Enclave* (QE), and Bob is the *MobileCoin Enclave*. Alice provides her target info via `aesmd` and the SGX SDK to the MobileCoin application, which then gives it to it's own enclave. Bob uses `EREPORT` to get his report, which he gives back to the QE.

The QE has (on startup) established a secure connection to Intel and gotten itself "provisioned" with a key it can use for communicating back to IAS. When it gets the report from Bob, the QE first authenticates the report (as Alice did above), then fills in extra details about the host: it's own version and details, the BIOS version, some security-relevant BIOS configuration details (e.g. "is hyperthreading turned on" or "is the GPU turned on", etc.), etc. and creates what's called a `Quote`.

In addition to the contents of Bob's `Quote` is the contents of Bob's `EREPORT`, details about the QE and the host, and additional information which is encrypted between the QE to Intel.

### IAS

> One the MobileCoin application has gotten a `Quote` back from the QE, it encodes it into a JSON document and sends it to the *Intel Attestation Service* (IAS). IAS, for it's part, looks at the quote and tries to verify everything: is the QE up-to-date? Does the CPU have any known vulnerabilities? IAS, then, is verifying that there aren't any known problems with the TCB---that you can actually T the CB.

Reviewer 2: ```suggestion
Once the MobileCoin application has gotten a `Quote` back from the QE, it encodes it into a JSON document and sends it to the *Intel Attestation Service* (IAS). IAS, for it's part, looks at the quote and tries to verify everything: is the QE up-to-date? Does the CPU have any known vulnerabilities? IAS, then, is verifying that there aren't any known problems with the TCB---that you can actually T the CB.
```

It puts all that (alongside "update info" if something is not up-to-date) into a report, and signs the bytes using an RSA key, then provides a standard certificate chain.

### Verification

> The software the MobileCoin enclave is attesting to (the client) must now verify the report provided by IAS. In particular, it must authenticate the message was signed by a key it trusts. Doing so is fairly straightforward RSA signature checking, where the verifier is given an X509 certificate chain and signature alongside the  authenticate that the chain leads back to a trust anchor the application expects.

Reviewer 2: ```suggestion
The software the MobileCoin enclave is attesting to (the client) must now verify the report provided by IAS. In particular, it must authenticate the message was signed by a key it trusts. Doing so is fairly straightforward RSA signature checking, where the verifier is given an X509 certificate chain and signature alongside the authenticate that the chain leads back to a trust anchor the application expects.
```

Because both Alice and Bob are using an asymmetric cryptosystem (where you can't figure out what private key is behind a public key), and neither one ever sends their private keys, the "new number" they come up with is a "shared secret". The "Eve" passing messages back and forth doesn't have either Alice's or Bob's private keys to multiply with the public key.

Bob and Alice now have a secret which each of them knows, but Eve cannot know, which means they can use this shared secret to create their own message authentication codes, between each other, without relying on special CPU instructions like `EREPORT`. When Bob wants to send a message to Alice, he can compute the authentication code for it (the MAC), and send both of them to Alice. Alice, for her part, will independently compute the MAC based on the message, and then compare her MAC with the one Bob sent over. If they match, then she knows that Eve didn't change the message.

> In other words, Alice can now authenticate messages from Bob, and visa-versa. If Alice and Bob also want to *hide* their messages from Eve, then Bob would first encrypt his message, then compute the MAC of the encrypted text. It's usually easier to both encrypt and MAC, because there is a special kind of encryption "Authenticated Encryption" that combines encryption and authentication codes into a single package.

Author: I think this might be better served by a separate section explaining that client-to-node communications work in basically the same way.

> In other words, Alice can now authenticate messages from Bob, and visa-versa. If Alice and Bob also want to *hide* their messages from Eve, then Bob would first encrypt his message, then compute the MAC of the encrypted text. It's usually easier to both encrypt and MAC, because there is a special kind of encryption "Authenticated Encryption" that combines encryption and authentication codes into a single package.

Reviewer 3: This section felt confusing because enclaves on the same machine aren't opening a communication channel with each other in practice. Better to reference the most 'up-to-date' concepts in the readers' minds - AVRs and client-to-enclave.

> In other words, Alice can now authenticate messages from Bob, and visa-versa. If Alice and Bob also want to *hide* their messages from Eve, then Bob would first encrypt his message, then compute the MAC of the encrypted text. It's usually easier to both encrypt and MAC, because there is a special kind of encryption "Authenticated Encryption" that combines encryption and authentication codes into a single package.

Author: OK, I'll take a stab at making this more clear.
